src/libopenrave-core/jsonparser/jsonwriter.cpp

```diff
@@ -374,23 +289,101 @@ void RaveWriteMsgPackMemory(EnvironmentBasePtr penv, std::vector<char>& output,
     OpenRAVE::MsgPack::DumpMsgPack(doc, output);
 }
 
-void RaveWriteMsgPackMemory(KinBodyPtr pbody, std::vector<char>& output, const AttributesList& atts, rapidjson::Document::AllocatorType& alloc)
+void RaveWriteMsgPackMemory(const std::list<KinBodyPtr>& listbodies, std::vector<char>& output, const AttributesList& atts, rapidjson::Document::AllocatorType& alloc)
 {
     rapidjson::Document doc(&alloc);
 
     EnvironmentJSONWriter jsonwriter(atts, doc, doc.GetAllocator());
-    jsonwriter.Write(pbody);
+    jsonwriter.Write(listbodies);
     OpenRAVE::MsgPack::DumpMsgPack(doc, output);
 }
 
-void RaveWriteMsgPackMemory(const std::list<KinBodyPtr>& listbodies, std::vector<char>& output, const AttributesList& atts, rapidjson::Document::AllocatorType& alloc)
+void RaveWriteEncryptedFile(EnvironmentBasePtr penv, const std::string& filename, const AttributesList& atts, rapidjson::Document::AllocatorType& alloc, MimeType mimeType)
 {
-    rapidjson::Document doc(&alloc);
+    std::ofstream ofs(filename, std::ios::binary);
+    RaveWriteEncryptedStream(penv, ofs, atts, alloc, mimeType);
+}
 
-    EnvironmentJSONWriter jsonwriter(atts, doc, doc.GetAllocator());
-    jsonwriter.Write(listbodies);
-    OpenRAVE::MsgPack::DumpMsgPack(doc, output);
+void RaveWriteEncryptedFile(const std::list<KinBodyPtr>& listbodies, const std::string& filename, const AttributesList& atts, rapidjson::Document::AllocatorType& alloc, MimeType mimeType)
+{
+    std::ofstream ofs(filename, std::ios::binary);
+    RaveWriteEncryptedStream(listbodies, ofs, atts, alloc, mimeType);
 }
 
+void RaveWriteEncryptedMemory(EnvironmentBasePtr penv, std::vector<char>& output, const AttributesList& atts, rapidjson::Document::AllocatorType& alloc, MimeType mimeType)
+{
+    using vectorstream = ::boost::interprocess::basic_vectorstream<std::vector<char>, std::char_traits<char>>;
+    vectorstream vs(std::ios::binary);
+    RaveWriteEncryptedStream(penv, vs, atts, alloc, mimeType);
+    vs.swap_vector(output);
 }
 
+void RaveWriteEncryptedMemory(const std::list<KinBodyPtr>& listbodies, std::vector<char>& output, const AttributesList& atts, rapidjson::Document::AllocatorType& alloc, MimeType mimeType)
+{
+    using vectorstream = ::boost::interprocess::basic_vectorstream<std::vector<char>, std::char_traits<char>>;
+    vectorstream vs(std::ios::binary);
+    RaveWriteEncryptedStream(listbodies, vs, atts, alloc, mimeType);
+    vs.swap_vector(output);
+}
+
+void RaveWriteEncryptedStream(EnvironmentBasePtr penv, std::ostream& os, const AttributesList& atts, rapidjson::Document::AllocatorType& alloc, MimeType mimeType)
+{
+    std::string keyName;
+    for (const std::pair<std::string, std::string>& attribute : atts) {
+        if (attribute.first == "openravekey") {
```

Comment:
Can we name it `gpgkeyid`? perhaps also have `gpghomedir`?

Reply:
Actually, these key ids should be repeatable? Since we should be able to encrypt for multiple recipients.

Reply:
done

Reply:
How about gpghomedir?

Reply:
Isn't that by environment variable GNUPGHOME as you have tested?

Reply:
maybe let's avoid requiring GNUPGHOME envvar

Reply:
It's slightly hard to inject the options into `GpgDecrypt` as it is used inside `JSONDownloaderScope` which doesn't really accept a list of options/attributes.

Reply:
@ziyan for gpghomedir, is it ok to rely on GNUPGHOME envvar instead? you asked for it, so we need your reply.

Reply:
It's best to have it instead of caller setting environment vars before calling openrave. Imagine your `penv->Load` code calling `setenv` first in your program.

Reply:
I know the design-wise it's good to have an option, but currently I don't have a good idea on how to inject the attributes into `JSONDownloaderScope` to be used for gpg decryption functions

Reply:
Ok, I see.

Just curious, how would you set home dir with gpgme?

Reply:
GNUPGHOME envvar (as we discussed)

https://www.gnupg.org/documentation/manuals/gnupg/GPG-Configuration.html

Reply:
Should be `GpgME::Context::setEngineHomeDirectory`

Reply:
@superfashi I'm confused. You said that passing the GpgME::Context::setEngineHomeDirectory argument from env.Load() is difficult.

Reply:
yeah it is difficult, unless you want to inject the logic into Load